### DISCUSSION TITLE
Proposal of new Attribute for null-state static analysis based on enum values

### DISCUSSION THREAD

**Discussion by xtofs** [dotnet/csharplang#6803] [Language Ideas]
The [static analysis around non-nullable reference types](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis) is able to understand in `if` branches and `switch` cases that an out parameter can or can not be null. Prominent use is [Dictionary.TryGetValue](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.trygetvalue?view=net-7.0). (docs doesn't show the MaybeNullWhenAttribute see [source](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs,1013) for actual declaration).
Currently the MaybeNullWhen attribute does only allow to distinguish between the two boolean values.

With the C#11 [generic attributes feature](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/generic-attributes) it should be possible to add the following attribute that can be used (should be used) by the compiler to distinguish between more than two outcomes of a `TryGetX` methods, method that are returning an enum.

```
[AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
public sealed class MaybeNullWhenNot<T> : Attribute where T : struct, Enum
{
    public MaybeNullWhenNot(T returnValue)
    {
        ReturnValue = returnValue;
    }

    public T ReturnValue { get; }
}
```

The compiler can use this to infer nullability for  a method like the following that has three out parameters and returns an enum value corresponding to each put parameter. 
   ```
 public NodeKind GetKind(
        [MaybeNullWhenNot<NodeKind>(NodeKind.None)] out Leaf leaf,
        [MaybeNullWhenNot<NodeKind>(NodeKind.Pair)] out Pair pair,
        [MaybeNullWhenNot<NodeKind>(NodeKind.Triple)] out Triple triple
)
```

This can then be used in a switch statement to "dispatch" between the three outcomes and the compiler should infer which of the out parameters are null and which not:

```
switch (node.GetKind(out var leaf, out var pair, out var triple))
{
    case NodeKind.None:
        // leaf not null
        // pair and triple maybe null
        break;
    case NodeKind.Pair:
        // pair  not null
        // leaf and triple maybe null
        break;
    case NodeKind.Triple:
        // triple  not null
        // leaf and pair maybe null
        break;
}
```

The name "MaybeNullWhenNot". seems a bit overwhelming and almost sounds like a double negation but follows the naming  of  other Attributes for null-state static analysis. It can be read as "not-null (as indicated in the parameter type) when it returns that value, null otherwise" or short (but technically incorrect) as "NotNullWhen"

If `GetKind` in the example above is a virtual method that is implemented on the three subclasses Leaf, Pair Triple, this is quite efficient since it does not require any runtime type information (aka reflection) and casting. It is only using a virtual method dispatch and is still being safe (in terms of non-nullable reference types ) and ergonomic.

**Comment by RikkiGibson** [comment]
Related: https://github.com/dotnet/roslyn/issues/48463

**Reply by xtofs** [reply]
That is unfortunate that https://github.com/dotnet/roslyn/issues/48463 was closed because a) proposals/issues are typically judged by their merits and not by their amount of change/work. The estimated work should determine when something gets done not if.  b) the associated issue is not going into detail why the complier has to track multiple states. As long as the Attributes value is a (non flags) enum and each parameters attribute is "disjunct", there is always only one state that is true.
